### ISSUE TITLE
bug for stream-level metrics in periodic eval

### ISSUE THREAD

**Issue by AntonioCarta** [ContinualAI/avalanche#1093]
periodic eval is broken in some subtle cases. Instead of calling `eval` on the entire stream, it calls it multiple times for each experience, breaking stream-level metrics.

Reproduce:
```
from avalanche.benchmarks import SplitMNIST
from avalanche.training import Naive
from avalanche.models import SimpleMLP

if __name__ == '__main__':
    benchmark = SplitMNIST(5)
    model = SimpleMLP()
    strat = Naive(model, None, eval_mb_size=512, eval_every=1)

    stream = benchmark.test_stream
    strat.train(stream[0], eval_streams=list(stream[:2]))
```

outputs:
```
-- >> Start of training phase << --
-- >> Start of eval phase << --
-- Starting eval on experience 0 (Task 0) from test stream --
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:05<00:00,  1.08s/it]
> Eval on experience 0 (Task 0) from test stream ended.
        Loss_Exp/eval_phase/test_stream/Task000/Exp000 = 2.3881
        Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 0.0033
-- >> End of eval phase << --
        Loss_Stream/eval_phase/test_stream/Task000 = 2.3881
        Top1_Acc_Stream/eval_phase/test_stream/Task000 = 0.0033
-- >> Start of eval phase << --
-- Starting eval on experience 1 (Task 0) from test stream --
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:01<00:00,  2.12it/s]
> Eval on experience 1 (Task 0) from test stream ended.
        Loss_Exp/eval_phase/test_stream/Task000/Exp001 = 2.3041
        Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp001 = 0.2129
-- >> End of eval phase << --
        Loss_Stream/eval_phase/test_stream/Task000 = 2.3041
        Top1_Acc_Stream/eval_phase/test_stream/Task000 = 0.2129
```

which means it's calling `eval` two times instead of a single one with the entire stream.

**Comment by WenjinW** [comment]
I encounter the same problem with MTSimpleMLP(). Is there any progress?

**Comment by AndreaCossu** [comment]
The call to `list` split the stream into a list of single experiences which are treated as separate streams with only one experience. 
Calling `[stream[:2]]` fixes it.

**Comment by AntonioCarta** [comment]
@AndreaCossu I think we need a more robust solution. We should try to normalize the `eval_streams` argument so that all the reasonable values work (`stream[0]`, `[stream[0]]`, `stream[:2]`, `[stream[:2]]`)

**Comment by AndreaCossu** [comment]
We can simply add this to the `train` call:
```python
if not isinstance(eval_streams, list):
    eval_streams = [eval_streams]
```
to cover the cases you mentioned.

The case mentioned in this issue can be solved only by forbidding the user to add an object of type experience in the `eval_streams` list. We can easily catch this and raise an error. Otherwise, if we want to leave the freedom, it is up to the user to provide a correct input.

**Comment by AntonioCarta** [comment]
I don't think that check is enough. For example `eval_streams=list(stream[:2])` still considers the experiences as separate streams. We have to check the stream name and benchmark instance of the experiences to check whether they come from different streams or the same stream.

**Comment by 3koozy** [comment]
Could this affect EarlyStoppingPlugin? because I am having strange results with it for a while now and didn't know why.
find a text log snippet below with the issue.
the strategy early stops at epoch 32 while it achives improved eval_loss at epoch 25 (eval_loss = 0.4676) and 
32 (eval_loss = 0.4382) and 
.
parameters:
`EarlyStoppingPlugin(patience=10, val_stream_name='test_stream', metric_name = "Loss_Stream", mode = "min", margin=-0.01)`

```
for train_task, test_task in zip(train_stream, test_stream):
      enc_strategy.train(train_task, num_workers=0, eval_streams=[test_task])
```

```
#define strategy:
    enc_strategy= SupervisedTemplate(
      params["model"], params["optimizer"], params["criterion"], train_mb_size=params["batch_size"], train_epochs=params["num_epochs"],
      eval_mb_size=params["batch_size"], device=device, evaluator= eval_plugin, plugins= plugins, eval_every = 1)

```
```
-- Starting training on experience 0 (Task 0) from train stream --
Epoch 24 ended.
	Loss_Epoch/train_phase/train_stream/Task000 = 0.2042
	Top1_Acc_Epoch/train_phase/train_stream/Task000 = 0.9951
	Topk_5_Acc_Epoch/train_phase/train_stream/Task000 = 1.0000
-- >> Start of eval phase << --
-- Starting eval on experience 0 (Task 0) from test stream --
> Eval on experience 0 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp000 = 0.4996
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 0.9153
	Topk_5_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 1.0000
-- >> End of eval phase << --
	Accuracy_On_Trained_Experiences/eval_phase/test_stream/Task000 = 0.9153
	Loss_Stream/eval_phase/test_stream/Task000 = 0.4996
	StreamForgetting/eval_phase/test_stream = 0.0000
	Top1_Acc_Stream/eval_phase/test_stream/Task000 = 0.9153
	Topk_5_Acc_Stream/eval_phase/test_stream/Task000 = 1.0000
Epoch 25 ended.
	Loss_Epoch/train_phase/train_stream/Task000 = 0.1706
	Top1_Acc_Epoch/train_phase/train_stream/Task000 = 0.9976
	Topk_5_Acc_Epoch/train_phase/train_stream/Task000 = 1.0000
-- >> Start of eval phase << --
-- Starting eval on experience 0 (Task 0) from test stream --
> Eval on experience 0 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp000 = 0.4676
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 0.9322
	Topk_5_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 1.0000
-- >> End of eval phase << --
	Accuracy_On_Trained_Experiences/eval_phase/test_stream/Task000 = 0.9322
	Loss_Stream/eval_phase/test_stream/Task000 = 0.4676
	StreamForgetting/eval_phase/test_stream = 0.0000
	Top1_Acc_Stream/eval_phase/test_stream/Task000 = 0.9322
	Topk_5_Acc_Stream/eval_phase/test_stream/Task000 = 1.0000
Epoch 26 ended.
	Loss_Epoch/train_phase/train_stream/Task000 = 0.1682
	Top1_Acc_Epoch/train_phase/train_stream/Task000 = 0.9927
	Topk_5_Acc_Epoch/train_phase/train_stream/Task000 = 1.0000
-- >> Start of eval phase << --
-- Starting eval on experience 0 (Task 0) from test stream --
> Eval on experience 0 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp000 = 0.5499
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 0.8475
	Topk_5_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 1.0000
-- >> End of eval phase << --
	Accuracy_On_Trained_Experiences/eval_phase/test_stream/Task000 = 0.8475
	Loss_Stream/eval_phase/test_stream/Task000 = 0.5499
	StreamForgetting/eval_phase/test_stream = 0.0000
	Top1_Acc_Stream/eval_phase/test_stream/Task000 = 0.8475
	Topk_5_Acc_Stream/eval_phase/test_stream/Task000 = 1.0000
Epoch 27 ended.
	Loss_Epoch/train_phase/train_stream/Task000 = 0.1542
	Top1_Acc_Epoch/train_phase/train_stream/Task000 = 0.9927
	Topk_5_Acc_Epoch/train_phase/train_stream/Task000 = 1.0000
-- >> Start of eval phase << --
-- Starting eval on experience 0 (Task 0) from test stream --
> Eval on experience 0 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp000 = 0.5051
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 0.8814
	Topk_5_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 1.0000
-- >> End of eval phase << --
	Accuracy_On_Trained_Experiences/eval_phase/test_stream/Task000 = 0.8814
	Loss_Stream/eval_phase/test_stream/Task000 = 0.5051
	StreamForgetting/eval_phase/test_stream = 0.0000
	Top1_Acc_Stream/eval_phase/test_stream/Task000 = 0.8814
	Topk_5_Acc_Stream/eval_phase/test_stream/Task000 = 1.0000
Epoch 28 ended.
	Loss_Epoch/train_phase/train_stream/Task000 = 0.1419
	Top1_Acc_Epoch/train_phase/train_stream/Task000 = 0.9951
	Topk_5_Acc_Epoch/train_phase/train_stream/Task000 = 1.0000
-- >> Start of eval phase << --
-- Starting eval on experience 0 (Task 0) from test stream --
> Eval on experience 0 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp000 = 0.4935
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 0.8814
	Topk_5_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 1.0000
-- >> End of eval phase << --
	Accuracy_On_Trained_Experiences/eval_phase/test_stream/Task000 = 0.8814
	Loss_Stream/eval_phase/test_stream/Task000 = 0.4935
	StreamForgetting/eval_phase/test_stream = 0.0000
	Top1_Acc_Stream/eval_phase/test_stream/Task000 = 0.8814
	Topk_5_Acc_Stream/eval_phase/test_stream/Task000 = 1.0000
Epoch 29 ended.
	Loss_Epoch/train_phase/train_stream/Task000 = 0.1694
	Top1_Acc_Epoch/train_phase/train_stream/Task000 = 0.9902
	Topk_5_Acc_Epoch/train_phase/train_stream/Task000 = 1.0000
-- >> Start of eval phase << --
-- Starting eval on experience 0 (Task 0) from test stream --
> Eval on experience 0 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp000 = 0.5383
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 0.8644
	Topk_5_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 1.0000
-- >> End of eval phase << --
	Accuracy_On_Trained_Experiences/eval_phase/test_stream/Task000 = 0.8644
	Loss_Stream/eval_phase/test_stream/Task000 = 0.5383
	StreamForgetting/eval_phase/test_stream = 0.0000
	Top1_Acc_Stream/eval_phase/test_stream/Task000 = 0.8644
	Topk_5_Acc_Stream/eval_phase/test_stream/Task000 = 1.0000
Epoch 30 ended.
	Loss_Epoch/train_phase/train_stream/Task000 = 0.1451
	Top1_Acc_Epoch/train_phase/train_stream/Task000 = 0.9976
	Topk_5_Acc_Epoch/train_phase/train_stream/Task000 = 1.0000
-- >> Start of eval phase << --
-- Starting eval on experience 0 (Task 0) from test stream --
> Eval on experience 0 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp000 = 0.4838
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 0.8983
	Topk_5_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 1.0000
-- >> End of eval phase << --
	Accuracy_On_Trained_Experiences/eval_phase/test_stream/Task000 = 0.8983
	Loss_Stream/eval_phase/test_stream/Task000 = 0.4838
	StreamForgetting/eval_phase/test_stream = 0.0000
	Top1_Acc_Stream/eval_phase/test_stream/Task000 = 0.8983
	Topk_5_Acc_Stream/eval_phase/test_stream/Task000 = 1.0000
Epoch 31 ended.
	Loss_Epoch/train_phase/train_stream/Task000 = 0.0970
	Top1_Acc_Epoch/train_phase/train_stream/Task000 = 0.9976
	Topk_5_Acc_Epoch/train_phase/train_stream/Task000 = 1.0000
-- >> Start of eval phase << --
-- Starting eval on experience 0 (Task 0) from test stream --
> Eval on experience 0 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp000 = 0.4822
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 0.8814
	Topk_5_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 1.0000
-- >> End of eval phase << --
	Accuracy_On_Trained_Experiences/eval_phase/test_stream/Task000 = 0.8814
	Loss_Stream/eval_phase/test_stream/Task000 = 0.4822
	StreamForgetting/eval_phase/test_stream = 0.0000
	Top1_Acc_Stream/eval_phase/test_stream/Task000 = 0.8814
	Topk_5_Acc_Stream/eval_phase/test_stream/Task000 = 1.0000
Epoch 32 ended.
	Loss_Epoch/train_phase/train_stream/Task000 = 0.1285
	Top1_Acc_Epoch/train_phase/train_stream/Task000 = 0.9902
	Topk_5_Acc_Epoch/train_phase/train_stream/Task000 = 1.0000
-- >> Start of eval phase << --
-- Starting eval on experience 0 (Task 0) from test stream --
> Eval on experience 0 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp000 = 0.4382
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 0.9153
	Topk_5_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 1.0000
-- >> End of eval phase << --
	Accuracy_On_Trained_Experiences/eval_phase/test_stream/Task000 = 0.9153
	Loss_Stream/eval_phase/test_stream/Task000 = 0.4382
	StreamForgetting/eval_phase/test_stream = 0.0000
	Top1_Acc_Stream/eval_phase/test_stream/Task000 = 0.9153
	Topk_5_Acc_Stream/eval_phase/test_stream/Task000 = 1.0000
-- >> End of training phase << --
-- >> Start of eval phase << --
-- Starting eval on experience 0 (Task 0) from test stream --
> Eval on experience 0 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp000 = 0.4373
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 0.8814
	Topk_5_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 1.0000
-- Starting eval on experience 1 (Task 0) from test stream --
> Eval on experience 1 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp001 = 15.3549
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp001 = 0.0000
	Topk_5_Acc_Exp/eval_phase/test_stream/Task000/Exp001 = 0.0000
-- Starting eval on experience 2 (Task 0) from test stream --
> Eval on experience 2 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp002 = 23.8548
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp002 = 0.0000
	Topk_5_Acc_Exp/eval_phase/test_stream/Task000/Exp002 = 0.0000
-- Starting eval on experience 3 (Task 0) from test stream --
> Eval on experience 3 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp003 = 7.7324
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp003 = 0.0000
	Topk_5_Acc_Exp/eval_phase/test_stream/Task000/Exp003 = 0.0000
-- Starting eval on experience 4 (Task 0) from test stream --
> Eval on experience 4 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp004 = 22.0577
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp004 = 0.0000
	Topk_5_Acc_Exp/eval_phase/test_stream/Task000/Exp004 = 0.1846
-- Starting eval on experience 5 (Task 0) from test stream --
> Eval on experience 5 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp005 = 9.2164
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp005 = 0.0000
	Topk_5_Acc_Exp/eval_phase/test_stream/Task000/Exp005 = 0.2241
-- Starting eval on experience 6 (Task 0) from test stream --
> Eval on experience 6 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp006 = 13.4966
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp006 = 0.0000
	Topk_5_Acc_Exp/eval_phase/test_stream/Task000/Exp006 = 0.0615
-- >> End of eval phase << --
	Accuracy_On_Trained_Experiences/eval_phase/test_stream/Task000 = 0.8814
	Loss_Stream/eval_phase/test_stream/Task000 = 13.1206
	StreamForgetting/eval_phase/test_stream = 0.0000
	Top1_Acc_Stream/eval_phase/test_stream/Task000 = 0.1238
	Topk_5_Acc_Stream/eval_phase/test_stream/Task000 = 0.2095
-- >> Start of training phase << --
-- >> Start of eval phase << --
-- Starting eval on experience 1 (Task 0) from test stream --
> Eval on experience 1 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp001 = 3.8279
-- >> End of eval phase << --
	Loss_Stream/eval_phase/test_stream/Task000 = 3.8279
-- Starting training on experience 1 (Task 0) from train stream --
```
